include/modem/pdn.h

```diff
@@ -0,0 +1,196 @@
+ * generate events from the +CNEC and +GGEV AT notifications to report
+ * state of the Packet Data Network (PDN) connection.
+ *
+ * @param[out] cid The ID of the new PDP context.
+ * @param cb Optional event handler.
+ * @return int Zero on success or a negative errno otherwise.
+ */
+int pdn_cid_create(uint8_t *cid, pdn_event_handler_t cb);
+
+/**
+ * @brief Configure a Packet Data Protocol context.
+ *
+ * @param cid The PDP context to configure.
+ * @param apn The Access Point Name to configure the PDP context with.
+ * @param family The family to configure the PDN context for.
+ * @param opts Optional additional configuration options.
+ * @return int Zero on success or a negative errno otherwise.
+ */
+int pdn_cid_configure(uint8_t cid, const char *apn, enum pdn_fam family,
+		      struct pdn_pdp_opt *opts);
+
+/**
+ * @brief Set PDN connection authentication parameters.
+ *
+ * @param cid The PDP context to set authentication parameters for.
+ * @param method The desired authentication method.
+ * @param user The username to use for authentication.
+ * @param password The password to use for authentication.
+ * @return int Zero on success or a negative errno otherwise.
+ */
+int pdn_cid_auth_set(uint8_t cid, enum pdn_auth method,
+		     const char *user, const char *password);
+
+/**
+ * @brief Destroy a Packet Data Protocol context.
+ *
+ * @param cid The PDP context to destroy.
+ * @return int Zero on success or a negative errno otherwise.
+ */
+int pdn_cid_destroy(uint8_t cid);
+
+/**
+ * @brief Activate a Packet Data Network (PDN) connection.
+ *
+ * @param cid The PDP context ID to activate a connection for.
+ * @param[out] esm If provided, the function will block to return
+ *		   the ESM error reason.
+ * @return int Zero on success or a negative errno otherwise.
+ */
+int pdn_activate(uint8_t cid, int *esm);
+
+/**
+ * @brief Deactivate a Packet Data Network (PDN) connection.
+ *
+ * @param cid The PDP context ID.
+ * @return int Zero on success or a negative errno otherwise.
+ */
+int pdn_deactivate(uint8_t cid);
+
+/**
+ * @brief Retrieve the PDN ID for a given PDP Context.
+ *
+ * The PDN ID can be used to route traffic through a Packet Data Network.
+ *
+ * @param cid The context ID of the PDN connection.
+ * @return int A non-negative PDN ID on success, or a negative errno otherwise.
+ */
+int pdn_id_get(uint8_t cid);
+
+/**
+ * @brief Retrieve the default Access Point Name (APN).
+ *
+ * The default APN is the APN of the default PDP context (zero).
+ *
+ * @param[out] buf The buffer to copy the APN into.
+ * @param len The size of the output buffer.
+ * @return int Zero on success or a negative errno otherwise.
+ */
+int pdn_default_apn_get(char *buf, size_t len);
+
+/**
+ * @brief Set a callback for events pertaining to the default PDP context (zero).
+ *
+ * @param cb The PDN event handler.
+ */
+int pdn_default_callback_set(pdn_event_handler_t cb);
+
```

Comment:
In order to browse contexts and their information, I'd like to have also functions like:
- pdn_context_info_get(cid)
-> i.e. to return the stuff from +CGDCONT / +CGCONTRDP / +CGACT: cid & pdn id, ipv4 address, ipv6 address, dns addresses, is it active or not
- pdn_context_info_listing()
- > similarly than previous but for all contexts

Reply:
I purposely did not add that, because `AT+CGDCONT?` output is large and I want to avoid having yet another library wrapping AT commands  that takes big chunks of RAM to parse an AT response. Some of that information can easily be maintained by the application, i.e. CID and APN are known by the application, state (active/inactive) is reported via events, PDN ID can be retrieved via API. The rest, I suggest that the application uses AT commands and parses the response manually.